Fix #2952: Add support for disabling practice tab

BenHenning opened the pull request:
## Explanation
Fix #2952

Introduces support for disabling the practice tab in the topic fragment. This is done via a new Dagger boolean qualified by ``EnablePracticeTab``. In order to easily disable the tab, ``TopicTab`` has been updated to actually compute/manage both tab positions and count based on whether the practice tab should be enabled.

Some other minor refactoring was done in ``TopicTab`` to try and consolidate tab-specific resources & strings into a single location which avoids extra ``when`` statements elsewhere.

I verified that TopicFragmentTest is also passing locally when run with Espresso on a Nexus 5X.

BenHenning committed:
Add support for loading proto versions of lessons.

BenHenning committed:
Upgrade Espresso versions to work around build issues.

Note that I can't actually verify Espresso tests are working with these
versions since the Bazel project doesn't yet have Espresso tests set up.

BenHenning committed:
Merge branch 'fix-protobuf-duplication-issue' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Lint fix post-copying from #2927.

BenHenning committed:
Fix Glide in Bazel.

BenHenning committed:
Fix concept card name escaping.

Also, refactor HtmlParser to better support/handling different tag types
& prepare for LaTeX SVG image support.

BenHenning committed:
Copy over test fix from #2927.

BenHenning committed:
Add support for rendering LaTeX via SVGs.

Much of this is copied from #2927 so see that PR's commit history & the
PR corresponding to this commit for specific details.

BenHenning committed:
Ensure all thumbnails are opaque.

BenHenning committed:
Fix interactions to support new localization.

This fixes answer classification, submission, and viewing for drag &
drop, item selection, multiple choice, and text input interactions per
changes in the states structure to support lesson localization.

BenHenning committed:
Fix blurring for SVGs.

SVGs did not previous support blurring, & this change introduces a way
to leverage the existing bitmap blurrer to introduce blurring
functionality in a way that's consistent with bitmap thumbnails.

BenHenning committed:
Lint fixes post copying from #2927.

BenHenning committed:
Remove image loading annotation.

This will be added in a later branch in the chain.

BenHenning committed:
Add caching module to needed test suites.

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning committed:
Merge branch 'fix-concept-card-name-rendering' into support-latex-svg-rendering

BenHenning committed:
Re-add image load annotation for this PR.

BenHenning committed:
Merge branch 'support-latex-svg-rendering' into fix-thumbnail-alpha-overrides

BenHenning committed:
Merge branch 'fix-thumbnail-alpha-overrides' into fix-locked-lesson-blurring

BenHenning committed:
Merge branch 'fix-locked-lesson-blurring' into add-support-for-item-selection-multiple-choice-drag-drop-localization

BenHenning committed:
Add support for hiding the questions player.

BenHenning committed:
Add missing dep from #2927 to fix build.

BenHenning commented:
/cc @anandwana001 @rt4914

BenHenning committed:
Add support for importing text protos as assets.

This actually provides Bazel-only support for converting text protos to
binary and including those as assets so that we can avoid checking in
binary protos to the codebase.

BenHenning committed:
Add gitignore for Android Studio Bazel plugin output.

BenHenning committed:
Import textproto versions of existing dev assets.

This imports local conversions of all existing dev assets as text proto
leveraging the Bazel-only conversion system to convert these to binary
protos and include them as assets.

This is a necessary step to both eventually replace JSON assets with
text proto, and to test that the new proto loading pipeline is working
as designed.

BenHenning committed:
Undo Espresso version change since it breaks Bazel tests.

BenHenning committed:
Fix tests.

Various attempts to fix existing tests + move some of the domain tests
over to actually using proto when running with Bazel.

App module tests are hanging currently for unknown reasons (only in
Bazel).

BenHenning committed:
Use correct Espresso core.

3.1.0 causes tests to hang indefinitely when run with Bazel.

rt4914 commented:
@BenHenning Does this PR still needs a review?

BenHenning committed:
Move coroutine dispatchers.

This moves the coroutine dispatchers to a new threading subpackage. It
also simplifies the TestCoroutineDispatcher interface & implementations
(which required moving runUntilIdle to CoroutineExecutorServiceTest).

The test coroutine annotations were also moved to their own files.

BenHenning committed:
Refactor testing utilities build graph.

This introduces new Bazel libraries for:
- Test dispatchers
- Fake system/Oppia clocks
- Robolectric dependencies

BenHenning committed:
Introduce dedicated test for TestCoroutineDispatcher.

This also fixes some threading issues in the dispatcher, and clarifies
some of its API that was previously unclear (and can lead to subtle race
conditions).

anandwana001 commented:
I will review this today.

BenHenning committed:
Add tests for TestCoroutineDispatcher.

This required a bunch of refactoring, adding the first examples of
dedicated Bazel test build files, and introducing a new test pattern for
sharing code between different implementations. Further, it refined the
API for TesCoroutineDispatcher & fixed some issues in the
implementations (especially Robolectric).

Tests verified as non-flaky in Robolectric across 1000 runs, and 0.2%
flaky across 1000 runs for Espresso (though that may be fixed; waiting
on a follow-up run to confirm stability).

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

anandwana001 reviewed (approved):
LGTM, Thanks @BenHenning

BenHenning committed:
Workaround ktlint semicolon issue.

BenHenning committed:
Lint fixes.

BenHenning committed:
Fix test post-merge.

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Move CoroutineDispatcher to be part of threading.

BenHenning committed:
Lint fixes.

anandwana001 commented:
https://github.com/oppia/oppia-android/pull/3072 blocked on this PR as #2986 has a lot of change around `topicTab` enum and TopicInfo which also requires here in this PR. 
@rt4914 Could you review this pr so we can merge it asap.

BenHenning committed:
Merge branch 'develop' into fix-protobuf-duplication-issue

BenHenning committed:
Undo version upgrade.

rt4914 commented on app/src/main/java/org/oppia/android/app/topic/revisioncard/PracticeTabModule.kt:
I think by mistake this file shifted to `revisioncard` package.

BenHenning committed:
Merge branch 'fix-protobuf-duplication-issue' into add-support-for-proto-lesson-loading-from-local-assets

Conflicts:
	.gitignore
	app/src/sharedTest/java/org/oppia/android/app/story/StoryActivityTest.kt

rt4914 reviewed (approved):
LGTM, one nit change and rest is blocked on tests.

BenHenning committed:
Fix broken tests.

BenHenning committed:
Lint fixes.

BenHenning committed:
Add a bit more proto loading test coverage.

BenHenning committed:
Comment fix.

BenHenning committed:
Lint fixes.

BenHenning committed:
Address earlier TODO.

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

Conflicts:
	tools/BUILD.bazel

BenHenning committed:
Post-merge restructure.

BenHenning committed:
Merge branch 'develop' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

Conflicts:
	data/BUILD.bazel

BenHenning committed:
Post-merge fixes.

This removes unneeded dependencies from PersistentCacheStore, including
Glide.

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Add tests for new tag handlers.

Also, add tests for HtmlParser, and clean up some existing tests.

BenHenning committed:
Fix broken test build.

BenHenning committed:
Merge branch 'add-coroutine-dispatcher-tests' into fix-glide-image-loading

Conflicts:
	app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicListControllerTest.kt
	testing/src/test/java/org/oppia/android/testing/story/StoryProgressTestHelperTest.kt

BenHenning committed:
Post-merge lint fixes.

BenHenning committed:
Merge branch 'add-coroutine-dispatcher-tests' into add-support-for-proto-lesson-loading-from-local-assets

Conflicts:
	app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicListControllerTest.kt
	testing/src/test/java/org/oppia/android/testing/story/StoryProgressTestHelperTest.kt

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

Conflicts:
	app/BUILD.bazel
	app/src/sharedTest/java/org/oppia/android/app/parser/HtmlParserTest.kt
	testing/BUILD.bazel
	utility/BUILD.bazel

BenHenning committed:
Lint fixes.

BenHenning committed:
Address reviewer comment.

BenHenning committed:
Post-merge fix.

BenHenning committed:
Merge branch 'fix-concept-card-name-rendering' into support-latex-svg-rendering

Conflicts:
	data/BUILD.bazel
	utility/src/main/java/org/oppia/android/util/caching/CacheAssetsLocally.kt

BenHenning committed:
Address pre-merge TODO.

BenHenning committed:
Disable local image loading.

Also, add specific LaTeX line in the first fractions interaction so that
the new SVG support can be tested.

BenHenning committed:
Fix typo.

BenHenning committed:
Merge branch 'fix-concept-card-name-rendering' into support-latex-svg-rendering

BenHenning committed:
Docs + tests.

This adds more documentation & context for code added before without it.
It also adds a bunch of new testing coverage for changes. The tests are
particularly around the html part of the LaTeX loading pipeline, but not
the image parts.

More of both will be needed specifically for the image loading pipeline
changes.

BenHenning committed:
Docs + tests + refactor.

This adds documentation for all remaining components introduced in #2981,
adds some testing for UrlImageParser, and cleans up the image loading
pipeline to be a bit simpler & cleaner (though this will probably break
down-stream PRs relying on image transformations).

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Merge branch 'add-coroutine-dispatcher-tests' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning committed:
Merge branch 'fix-concept-card-name-rendering' into support-latex-svg-rendering

BenHenning committed:
Lint fixes.

BenHenning committed:
Fix lint failure & re-enable assets.

BenHenning committed:
Merge branch 'develop' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning committed:
Merge branch 'fix-concept-card-name-rendering' into support-latex-svg-rendering

BenHenning committed:
Merge branch 'support-latex-svg-rendering' into fix-thumbnail-alpha-overrides

BenHenning committed:
Clean up + tests.

Add tests where reasonable given current testing limitations and
developing time availability, and remove some dead code. Add new testing
utility (GenericViewMatchers) + corresponding Bazel package.

BenHenning committed:
Consolidate testing module view matchers.

This moves two other view matchers that are part of the testing module
into the new Espresso package.

BenHenning committed:
Lint fixes.

BenHenning committed:
Merge branch 'fix-thumbnail-alpha-overrides' into fix-locked-lesson-blurring

Conflicts:
	utility/src/main/java/org/oppia/android/util/parser/SvgPictureDrawable.kt

BenHenning committed:
Post-merge fixes + docs.

Add TODOs for tests that can't easily be added. Fill in missing
documentation. Fix post-merge issues to ensure SVGs can actually be
blurred (will be verified after this commit).

BenHenning commented:
Ack thanks for the notice of blocking @anandwana001. I'm working to get all upstream work this depends on in. Nearly everything before this PR is now in review (only #2985 isn't yet).

BenHenning committed:
Merge branch 'fix-locked-lesson-blurring' into add-support-for-item-selection-multiple-choice-drag-drop-localization

Conflicts:
	domain/src/main/java/org/oppia/android/domain/util/InteractionObjectExtensions.kt

BenHenning committed:
Add tests for structural changes.

Revamps StateFragmentTest to include a bunch of new interaction tests to
verify that the new structural changes are correct. These discovered
issues with the old asset converter that needed to be fixed, and new
textproto changes needed to be pulled in.

Removed a few now-unused (or should be unused) asset files.
ExplorationProgressControllerTest is now broken and has to be almost
entirely rewritten to support different explorations. The previous ones
couldn't reasonably be migrated without lots of manual work, so we
abandoned them.

BenHenning committed:
Rename test (& fix it).

Relevant fixes will be coming in a later commit. This is just committing
the file move.

BenHenning committed:
Fix/update json assets.

This includes fixes for existing content in textproto files, too, where
the previous conversion script had issues.

Assets that will no longer be included are removed (due to being too
hard to migrate).

The codebase will be broken at this point--thorough test changes are
needed, first.

BenHenning committed:
Fix a LOT of tests that break due to changes.

Fundamentally, this:
1. Migrates existing structural changes in tests for item selection,
text input, and drag and drop.
2. Introduces proto loading (when run on Bazel) for a few more tests,
including some UI tests.
3. Updates test suites to no longer depend on explorations that weren't
carried over in this change.
4. Cleans up a bunch of existing tests/line wrapping issues and other
things.

BenHenning committed:
Import structurally updates questions.

This includes proto questions even though they can't yet be imported.
Some tests needed adjustments due to fixes in the questions themselves.

BenHenning committed:
Merge branch 'develop' into fix-glide-image-loading

BenHenning committed:
Merge branch 'fix-glide-image-loading' into fix-concept-card-name-rendering

BenHenning committed:
Merge branch 'add-support-for-item-selection-multiple-choice-drag-drop-localization' into add-support-for-disabling-practice-tab

Conflicts:
	app/src/main/java/org/oppia/android/app/application/ApplicationComponent.kt
	app/src/sharedTest/java/org/oppia/android/app/options/OptionsFragmentTest.kt
	app/src/sharedTest/java/org/oppia/android/app/topic/practice/TopicPracticeFragmentTest.kt

BenHenning committed:
Merge branch 'develop' into fix-concept-card-name-rendering

BenHenning committed:
Merge branch 'fix-concept-card-name-rendering' into support-latex-svg-rendering

BenHenning committed:
Fix post-merge test build failures.

BenHenning committed:
Merge branch 'develop' into support-latex-svg-rendering

BenHenning committed:
Merge branch 'support-latex-svg-rendering' into fix-thumbnail-alpha-overrides

BenHenning committed:
Add test for when practice tab is disabled.

BenHenning commented on app/src/main/java/org/oppia/android/app/topic/revisioncard/PracticeTabModule.kt:
Nice catch. Fixed (will be visible in the next push to this branch).

BenHenning committed:
Fix new module location per reviewer comment.

BenHenning committed:
Merge branch 'develop' into fix-thumbnail-alpha-overrides

BenHenning committed:
Merge branch 'fix-thumbnail-alpha-overrides' into fix-locked-lesson-blurring

BenHenning committed:
Merge branch 'develop' into fix-locked-lesson-blurring

BenHenning committed:
Merge branch 'fix-locked-lesson-blurring' into add-support-for-item-selection-multiple-choice-drag-drop-localization

BenHenning committed:
Post-merge lint fixes.

BenHenning committed:
Address reviewer comments.

Fix case sensitive checks for text input, and some minor cleanups in
drag & drop interaction view model.

BenHenning committed:
Fix broken Gradle test.

Another structural change: 'translation' is now the correct reference
for written translations rather than 'html'.

BenHenning committed:
Merge branch 'add-support-for-item-selection-multiple-choice-drag-drop-localization' into add-support-for-disabling-practice-tab

BenHenning committed:
Lint fixes.

BenHenning commented:
Thanks @anandwana001 & @rt4914 for the reviews.

@rt4914 given some new modules were migrated, your earlier comment, and the new test to verify the tab properly hides, PTAL again when you get a chance.

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/topic/TopicFragmentTest.kt:
I think there should be tests to check that on configuration-change the tab remains hidden/visible.

rt4914 reviewed (approved):
@BenHenning Overall LGTM. Suggested to add one test case. 
Not blocking this PR on that test suggestion.

BenHenning committed:
Merge branch 'develop' into add-support-for-item-selection-multiple-choice-drag-drop-localization

BenHenning committed:
Merge branch 'add-support-for-item-selection-multiple-choice-drag-drop-localization' into add-support-for-disabling-practice-tab

BenHenning commented:
Only Bazel failure was RevisionCardActivityLocalTest which seems to be due to a network flake (a Maven dependency failed to download). I think we can safely ignore this and avoid needing to re-run Bazel tests before merging this PR in.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/topic/TopicFragmentTest.kt:
Nice idea. Done.

BenHenning committed:
Address reviewer comment.

Adds test with config change.

BenHenning commented:
Thanks for the review! I added the suggest test change & verified that TopicFragmentTest is passing locally with Espresso.

Note I'll probably cancel the Bazel tests per my comment above & verifying tests passing locally. I'd prefer to get the whole chain in quickly since I'll be verifying that Bazel is passing at the end (after #3116 is merged).
